File: tests/go/benchmark.go
//go:build ignore
// +build ignore

package main

import (
	"encoding/json"
	"fmt"
	"runtime"
	"sort"
	"time"

	"github.com/aptos-labs/aptos-go-sdk"
	"github.com/aptos-labs/aptos-go-sdk/crypto"
	"golang.org/x/crypto/sha3"
)

type BenchmarkResult struct {
	Name string  `json:"name"`
	Avg  float64 `json:"avg"`
	P95  float64 `json:"p95"`
	Min  float64 `json:"min"`
	Max  float64 `json:"max"`
	Ops  float64 `json:"ops"`
	Unit string  `json:"unit"`
}

func calculateStats(samples []float64) (avg, p95, min, max float64) {
	if len(samples) == 0 {
		return 0, 0, 0, 0
	}

	sorted := make([]float64, len(samples))
	copy(sorted, samples)
	sort.Float64s(sorted)

	var sum float64
	for _, s := range samples {
		sum += s
	}
	avg = sum / float64(len(samples))

	p95Idx := int(float64(len(sorted)) * 0.95)
	if p95Idx >= len(sorted) {
		p95Idx = len(sorted) - 1
	}
	p95 = sorted[p95Idx]
	min = sorted[0]
	max = sorted[len(sorted)-1]

	return avg, p95, min, max
}

func measureSync(fn func(), iterations int, warmup int) (avg, p95, min, max float64) {
	// Warmup
	for i := 0; i < warmup; i++ {
		fn()
	}

	// Measure
	samples := make([]float64, iterations)
	for i := 0; i < iterations; i++ {
		start := time.Now()
		fn()
		elapsed := float64(time.Since(start).Nanoseconds()) / 1000.0 // microseconds
		samples[i] = elapsed
	}

	return calculateStats(samples)
}

func measureAsync(fn func() error, iterations int, warmup int) (avg, p95, min, max float64, err error) {
	// Warmup
	for i := 0; i < warmup; i++ {
		if err := fn(); err != nil {
			return 0, 0, 0, 0, err
		}
	}

	// Measure
	samples := make([]float64, iterations)
	for i := 0; i < iterations; i++ {
		start := time.Now()
		if err := fn(); err != nil {
			return 0, 0, 0, 0, err
		}
		elapsed := float64(time.Since(start).Milliseconds()) // milliseconds
		samples[i] = elapsed
	}

	avg, p95, min, max = calculateStats(samples)
	return avg, p95, min, max, nil
}

func main() {
	fmt.Println("=== Aptos Go SDK Benchmarks ===")
	fmt.Println()
	fmt.Printf("Network: Devnet\n")
	fmt.Printf("Runtime: Go %s\n", runtime.Version())
	fmt.Printf("Date: %s\n\n", time.Now().Format(time.RFC3339))

	results := []BenchmarkResult{}

	// Initialize client
	client, err := aptos.NewClient(aptos.DevnetConfig)
	if err != nil {
		fmt.Printf("Failed to create client: %v\n", err)
		return
	}

	knownAddress := aptos.AccountOne

	// =============================================================================
	// Cryptographic Operations (Local)
	// =============================================================================
	fmt.Println("--- Cryptographic Operations ---")
	fmt.Println()

	// Ed25519 Key Generation
	{
		avg, p95, min, max := measureSync(func() {
			crypto.GenerateEd25519PrivateKey()
		}, 1000, 100)
		ops := 1000000.0 / avg
		results = append(results, BenchmarkResult{Name: "Ed25519 Key Gen", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "μs"})
		fmt.Printf("Ed25519 Key Gen: %.1f μs avg, %.0f ops/s\n", avg, ops)
	}

	// Ed25519 Signing
	{
		pk, _ := crypto.GenerateEd25519PrivateKey()
		message := make([]byte, 256)
		for i := range message {
			message[i] = 0x42
		}
		avg, p95, min, max := measureSync(func() {
			pk.SignMessage(message)
		}, 1000, 100)
		ops := 1000000.0 / avg
		results = append(results, BenchmarkResult{Name: "Ed25519 Sign", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "μs"})
		fmt.Printf("Ed25519 Sign: %.1f μs avg, %.0f ops/s\n", avg, ops)
	}

	// Ed25519 Verification
	{
		pk, _ := crypto.GenerateEd25519PrivateKey()
		pubKey := pk.PubKey().(*crypto.Ed25519PublicKey)
		message := make([]byte, 256)
		for i := range message {
			message[i] = 0x42
		}
		sig, _ := pk.SignMessage(message)
		ed25519Sig := sig.(*crypto.Ed25519Signature)
		avg, p95, min, max := measureSync(func() {
			pubKey.Verify(message, ed25519Sig)
		}, 1000, 100)
		ops := 1000000.0 / avg
		results = append(results, BenchmarkResult{Name: "Ed25519 Verify", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "μs"})
		fmt.Printf("Ed25519 Verify: %.1f μs avg, %.0f ops/s\n", avg, ops)
	}

	// SHA3-256 Hashing
	{
		message := make([]byte, 256)
		for i := range message {
			message[i] = 0x42
		}
		avg, p95, min, max := measureSync(func() {
			h := sha3.New256()
			h.Write(message)
			h.Sum(nil)
		}, 5000, 100)
		ops := 1000000.0 / avg
		results = append(results, BenchmarkResult{Name: "SHA3-256", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "μs"})
		fmt.Printf("SHA3-256 Hash: %.2f μs avg, %.0f ops/s\n", avg, ops)
	}

	// =============================================================================
	// REST API Operations (Network)
	// =============================================================================
	fmt.Println()
	fmt.Println("--- REST API Operations ---")
	fmt.Println()

	// Get Ledger Info
	{
		avg, p95, min, max, err := measureAsync(func() error {
			_, err := client.Info()
			return err
		}, 10, 3)
		if err != nil {
			fmt.Printf("Get Ledger Info: ERROR (%v)\n", err)
		} else {
			ops := 1000.0 / avg
			results = append(results, BenchmarkResult{Name: "Get Ledger Info", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "ms"})
			fmt.Printf("Get Ledger Info: %.1f ms avg, %.1f req/s\n", avg, ops)
		}
	}

	// Get Account Info
	{
		avg, p95, min, max, err := measureAsync(func() error {
			_, err := client.Account(knownAddress)
			return err
		}, 10, 3)
		if err != nil {
			fmt.Printf("Get Account Info: ERROR (%v)\n", err)
		} else {
			ops := 1000.0 / avg
			results = append(results, BenchmarkResult{Name: "Get Account Info", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "ms"})
			fmt.Printf("Get Account Info: %.1f ms avg, %.1f req/s\n", avg, ops)
		}
	}

	// Get Account Resources
	{
		avg, p95, min, max, err := measureAsync(func() error {
			_, err := client.AccountResources(knownAddress)
			return err
		}, 10, 3)
		if err != nil {
			fmt.Printf("Get Account Resources: ERROR (%v)\n", err)
		} else {
			ops := 1000.0 / avg
			results = append(results, BenchmarkResult{Name: "Get Account Resources", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "ms"})
			fmt.Printf("Get Account Resources: %.1f ms avg, %.1f req/s\n", avg, ops)
		}
	}

	// Get Account Balance
	{
		avg, p95, min, max, err := measureAsync(func() error {
			_, err := client.AccountAPTBalance(knownAddress)
			return err
		}, 10, 3)
		if err != nil {
			fmt.Printf("Get Account Balance: ERROR (%v)\n", err)
		} else {
			ops := 1000.0 / avg
			results = append(results, BenchmarkResult{Name: "Get Account Balance", Avg: avg, P95: p95, Min: min, Max: max, Ops: ops, Unit: "ms"})
			fmt.Printf("Get Account Balance: %.1f ms avg, %.1f req/s\n", avg, ops)
		}
	}

	// =============================================================================
	// Summary JSON
	// =============================================================================
	fmt.Println()
	fmt.Println("=== Results JSON ===")

	resultsMap := make(map[string]interface{})
	for _, r := range results {
		resultsMap[r.Name] = map[string]interface{}{
			"avg":  r.Avg,
			"p95":  r.P95,
			"ops":  r.Ops,
			"unit": r.Unit,
		}
	}

	output := map[string]interface{}{
		"sdk":       "go",
		"version":   "1.3.0",
		"runtime":   runtime.Version(),
		"network":   "devnet",
		"timestamp": time.Now().Format(time.RFC3339),
		"results":   resultsMap,
	}

	jsonBytes, _ := json.MarshalIndent(output, "", "  ")
	fmt.Println(string(jsonBytes))
}

File: tests/typescript/benchmark.ts
#!/usr/bin/env bun

/**
 * Standalone performance benchmark script for Aptos TypeScript SDK
 * Run with: bun benchmark.ts
 */

import {
  Aptos,
  AptosConfig,
  Network,
  Account,
} from "@aptos-labs/ts-sdk";
import { sha3_256 } from "@noble/hashes/sha3.js";

interface BenchmarkResult {
  name: string;
  avg: number;
  p95: number;
  min: number;
  max: number;
  ops: number;
  unit: string;
}

function calculateStats(samples: number[]): { avg: number; p95: number; min: number; max: number } {
  if (samples.length === 0) return { avg: 0, p95: 0, min: 0, max: 0 };
  
  const sorted = [...samples].sort((a, b) => a - b);
  const avg = samples.reduce((a, b) => a + b, 0) / samples.length;
  const p95Idx = Math.floor(samples.length * 0.95);
  const p95 = sorted[Math.min(p95Idx, sorted.length - 1)];
  const min = sorted[0];
  const max = sorted[sorted.length - 1];
  
  return { avg, p95, min, max };
}

async function measureAsync<T>(
  fn: () => Promise<T>,
  iterations: number,
  warmup: number = 3
): Promise<{ avg: number; p95: number; min: number; max: number }> {
  // Warmup
  for (let i = 0; i < warmup; i++) {
    await fn();
  }
  
  // Measure
  const samples: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    await fn();
    const end = performance.now();
    samples.push(end - start);
  }
  
  return calculateStats(samples);
}

function measureSync<T>(
  fn: () => T,
  iterations: number,
  warmup: number = 100
): { avg: number; p95: number; min: number; max: number } {
  // Warmup
  for (let i = 0; i < warmup; i++) {
    fn();
  }
  
  // Measure
  const samples: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    const end = performance.now();
    samples.push(end - start);
  }
  
  return calculateStats(samples);
}

async function runBenchmarks() {
  console.log("=== Aptos TypeScript SDK Benchmarks ===\n");
  console.log(`Network: Devnet`);
  console.log(`Runtime: Bun ${Bun.version}`);
  console.log(`Date: ${new Date().toISOString()}\n`);
  
  const results: BenchmarkResult[] = [];
  
  // Initialize client
  const config = new AptosConfig({ network: Network.DEVNET });
  const aptos = new Aptos(config);
  
  // Get a known address for tests
  const knownAddress = "0x1";
  
  // =============================================================================
  // Cryptographic Operations (Local)
  // =============================================================================
  console.log("--- Cryptographic Operations ---\n");
  
  // Ed25519 Key Generation
  {
    const stats = measureSync(() => Account.generate(), 1000);
    const avgUs = stats.avg * 1000;
    const ops = 1000000 / avgUs;
    results.push({ name: "Ed25519 Key Gen", avg: avgUs, p95: stats.p95 * 1000, min: stats.min * 1000, max: stats.max * 1000, ops, unit: "μs" });
    console.log(`Ed25519 Key Gen: ${avgUs.toFixed(1)} μs avg, ${ops.toFixed(0)} ops/s`);
  }
  
  // Ed25519 Signing
  {
    const account = Account.generate();
    const message = new Uint8Array(256).fill(0x42);
    const stats = measureSync(() => account.sign(message), 1000);
    const avgUs = stats.avg * 1000;
    const ops = 1000000 / avgUs;
    results.push({ name: "Ed25519 Sign", avg: avgUs, p95: stats.p95 * 1000, min: stats.min * 1000, max: stats.max * 1000, ops, unit: "μs" });
    console.log(`Ed25519 Sign: ${avgUs.toFixed(1)} μs avg, ${ops.toFixed(0)} ops/s`);
  }
  
  // Ed25519 Verification
  {
    const account = Account.generate();
    const message = new Uint8Array(256).fill(0x42);
    const signature = account.sign(message);
    const publicKey = account.publicKey;
    const stats = measureSync(() => publicKey.verifySignature({ message, signature }), 1000);
    const avgUs = stats.avg * 1000;
    const ops = 1000000 / avgUs;
    results.push({ name: "Ed25519 Verify", avg: avgUs, p95: stats.p95 * 1000, min: stats.min * 1000, max: stats.max * 1000, ops, unit: "μs" });
    console.log(`Ed25519 Verify: ${avgUs.toFixed(1)} μs avg, ${ops.toFixed(0)} ops/s`);
  }
  
  // SHA3-256 Hashing
  {
    const message = new Uint8Array(256).fill(0x42);
    const stats = measureSync(() => sha3_256(message), 5000);
    const avgUs = stats.avg * 1000;
    const ops = 1000000 / avgUs;
    results.push({ name: "SHA3-256", avg: avgUs, p95: stats.p95 * 1000, min: stats.min * 1000, max: stats.max * 1000, ops, unit: "μs" });
    console.log(`SHA3-256 Hash: ${avgUs.toFixed(2)} μs avg, ${ops.toFixed(0)} ops/s`);
  }
  
  // =============================================================================
  // REST API Operations (Network)
  // =============================================================================
  console.log("\n--- REST API Operations ---\n");
  
  // Get Ledger Info
  {
    const stats = await measureAsync(() => aptos.getLedgerInfo(), 5, 1);
    results.push({ name: "Get Ledger Info", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
    console.log(`Get Ledger Info: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
  }
  
  // Get Account Info
  {
    const stats = await measureAsync(() => aptos.getAccountInfo({ accountAddress: knownAddress }), 5, 1);
    results.push({ name: "Get Account Info", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
    console.log(`Get Account Info: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
  }
  
  // Get Account Resources
  {
    const stats = await measureAsync(() => aptos.getAccountResources({ accountAddress: knownAddress }), 5, 1);
    results.push({ name: "Get Account Resources", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
    console.log(`Get Account Resources: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
  }
  
  // Get Account Balance
  {
    const stats = await measureAsync(() => aptos.getAccountAPTAmount({ accountAddress: knownAddress }), 5, 1);
    results.push({ name: "Get Account Balance", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
    console.log(`Get Account Balance: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
  }
  
  // =============================================================================
  // GraphQL/Indexer Operations (Network)
  // =============================================================================
  console.log("\n--- GraphQL/Indexer Operations ---\n");
  
  // Get Account Tokens
  {
    try {
      const stats = await measureAsync(() => aptos.getAccountOwnedTokens({ accountAddress: knownAddress }), 10);
      results.push({ name: "Get Account Tokens", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
      console.log(`Get Account Tokens: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
    } catch (e) {
      console.log(`Get Account Tokens: SKIPPED (${e.message})`);
    }
  }
  
  // Get Account Transactions
  {
    try {
      const stats = await measureAsync(() => aptos.getAccountTransactions({ accountAddress: knownAddress, options: { limit: 10 } }), 10);
      results.push({ name: "Get Account Transactions", avg: stats.avg, p95: stats.p95, min: stats.min, max: stats.max, ops: 1000 / stats.avg, unit: "ms" });
      console.log(`Get Account Transactions: ${stats.avg.toFixed(1)} ms avg, ${(1000 / stats.avg).toFixed(1)} req/s`);
    } catch (e) {
      console.log(`Get Account Transactions: SKIPPED (${e.message})`);
    }
  }
  
  // =============================================================================
  // Summary JSON
  // =============================================================================
  console.log("\n=== Results JSON ===");
  console.log(JSON.stringify({
    sdk: "typescript",
    version: "5.2.0",
    runtime: `bun ${Bun.version}`,
    network: "devnet",
    timestamp: new Date().toISOString(),
    results: results.reduce((acc, r) => {
      acc[r.name] = { avg: r.avg, p95: r.p95, ops: r.ops, unit: r.unit };
      return acc;
    }, {} as Record<string, any>)
  }, null, 2));
}

runBenchmarks().catch(console.error);
